src/app/components/AdaptiveTrimmer/AdaptiveTrimmer.tsx

```diff
@@ -0,0 +1,140 @@
+  // Initial setup
+  const fullLength = text.length
+  const textRef = useRef<HTMLDivElement | null>(null)
+
+  // Data about the currently rendered version
+  const [currentContent, setCurrentContent] = useState('')
+  const [currentLength, setCurrentLength] = useState(0)
+
+  // Known good - this fits
+  const [largestKnownGood, setLargestKnownGood] = useState(0)
+
+  // Known bad - this doesn't fit
+  const [smallestKnownBad, setSmallestKnownBad] = useState(fullLength + 1)
+
+  // Are we exploring our possibilities now?
+  const [inDiscovery, setInDiscovery] = useState(false)
+
+  const attemptContent = useCallback((content: string, length: number) => {
+    setCurrentContent(content)
+    setCurrentLength(length)
+  }, [])
+
+  const attemptShortenedContent = useCallback(
+    (length: number) => {
+      const content =
+        strategy === 'middle'
+          ? trimLongString(text, Math.floor(length / 2) - 1, Math.floor(length / 2) - 1)!
+          : trimLongString(text, length, 0)!
+
+      attemptContent(content, length)
+    },
+    [strategy, text, attemptContent],
+  )
+
+  const initDiscovery = useCallback(() => {
+    setLargestKnownGood(0)
+    setSmallestKnownBad(fullLength + 1)
+    attemptContent(text, fullLength)
+    setInDiscovery(true)
+  }, [text, fullLength, attemptContent])
+
+  useEffect(() => {
+    initDiscovery()
+    const handleResize = () => {
+      initDiscovery()
+    }
+
+    window.addEventListener('resize', handleResize)
+    return () => window.removeEventListener('resize', handleResize)
+  }, [initDiscovery])
+
+  useEffect(() => {
+    if (inDiscovery) {
+      if (!textRef.current) {
+        return
+      }
+      const isOverflow = textRef.current.scrollWidth > textRef.current.clientWidth
+
+      if (isOverflow) {
+        // This is too much
+
+        // Update known bad length
+        const newSmallestKnownBad = Math.min(currentLength, smallestKnownBad)
+        setSmallestKnownBad(newSmallestKnownBad)
+
+        // We should try something smaller
+        attemptShortenedContent(Math.floor((largestKnownGood + newSmallestKnownBad) / 2))
+      } else {
+        // This is OK
+
+        // Update known good length
+        const newLargestKnownGood = Math.max(currentLength, largestKnownGood)
+        setLargestKnownGood(currentLength)
+
+        if (currentLength === fullLength) {
+          // The whole thing fits, so we are good.
+          setInDiscovery(false)
+        } else {
+          if (currentLength + 1 === smallestKnownBad) {
+            // This the best we can do, for now
+            setInDiscovery(false)
+          } else {
+            // So far, so good, but we should try something longer
+            attemptShortenedContent(Math.floor((newLargestKnownGood + smallestKnownBad) / 2))
+          }
+        }
+      }
+    }
+  }, [inDiscovery, currentLength, largestKnownGood, smallestKnownBad, attemptShortenedContent, fullLength])
```

Comment:
Too much complexity for a small improvement

Reply:
If we want this, please implement it using CSS. That will also fix <kbd>CTRL+F</kbd> "0x123456789" finding "0x123...789"

e.g. https://github.com/oasisprotocol/oasis-wallet-web/compare/master...lw/abandoned/address-ellipsis

Reply:
A pure CSS solution would be tempting, if it can fulfill all the requirements. Things to check:

- Can we instruct CSS to shorten the string by adding eclipses either in the middle or the end, based on our specific preference?
- If CSS is doing the shortening, then how do we know that we need to add a tooltip?

Reply:
- yes
- no

Reply:
Can the CSS solution do something similar to [this](https://a193f9df.oasis-explorer.pages.dev/testnet/pontusx/search?q=kunst)?

(When opened with narrow horizontal size, the name will be dynamically resized, beginning and end hidden, but it will always be displayed in a way that the highlight is within the visible section, even if it's not at the middle.)

Reply:
OK, going with css.